KAFKA-9633: Ensure ConfigProviders are closed

tombentley opened the pull request:
Fixes KAFKA-9633.

tombentley committed:
Ensure ConfigProviders are closed

Fixes KAFKA-9633.

tombentley commented:
Any chance of code reivew @rajinisivaram, @kkonstantine, @hachikuji?

kkonstantine commented:
ok to test

tombentley commented:
@kkonstantine any change of merging this?

tombentley commented:
@kkonstantine any change of merging this?

kkonstantine commented:
retest this please

kkonstantine commented:
ok to test

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -220,6 +220,8 @@ public void stop() {
         workerMetricsGroup.close();
         connectorStatusMetricsGroup.close();
+
+        workerConfigTransformer.close();
```
Looking at the initialization of `workerConfigTransformer` I see it should be made final. 

And then I notice that this is the case for `connectorClientConfigOverridePolicy` and all the class members of `ConnectorStatusMetricsGroup`. @tombentley do you mind tightening these types as well?

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfigTransformer.java:
```diff
@@ -98,4 +101,8 @@ public void onCompletion(Throwable error, Void result) {
         connectorRequests.put(path, request);
     }
+
+    public void close() {
```
should we also change this class to implement `AutoCloseable`? 
This can't be used immediately in a try-with-resources clause, but probably better to signal the existence of this method at the class level.

kkonstantine reviewed (commented):
Thanks @tombentley 
This seems almost ready. Added a couple minor comments that I noticed after another pass. 
Lmk what you think

kkonstantine commented:
JDK8 build failed on a relevant test: 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/2007/testReport/junit/org.apache.kafka.connect.runtime/WorkerTest/testStartAndStopConnector/

tombentley committed:
Review comment: make some fields final

tombentley committed:
Review comment: WorkerConfigTransformer implements AutoCloseable

tombentley committed:
Make MockFileConfigProvider.assertClosed() work multi-threaded

Gradle executes test classes concurrently, so MockFileConfigProvider
can't simply use a static field to hold its closure state.
Instead use a protocol whereby the MockFileConfigProvider is configured
with some unique ket identifying the test which also used when calling
assertClosed().

tombentley commented:
@kkonstantine both those suggestions seemed reasonable, so I made those changes. I've also fixed the problem with the test which was due to the tests running concurrently but using a `static` variable to make the assertion (since the test code can't access the mock provider instance). Please take another look.

kkonstantine commented on clients/src/test/java/org/apache/kafka/common/config/provider/MockFileConfigProvider.java:
```diff
@@ -19,11 +19,46 @@
+    public static void assertClosed(String id) {
+        MockFileConfigProvider instance = INSTANCES.remove(id);
+        assertNotNull(instance);
+        synchronized (instance) {
```
Seems `volatile` would works as well in this case. But that's fine for this test class.

kkonstantine reviewed (approved):
Thanks @tombentley !
Glad we caught the race condition in the tests. 

LGTM
Starting another hopefully final build.

kkonstantine commented:
retest this please

kkonstantine commented:
@tombentley there are a couple test cases that need adjustment after the recent fix on the race condition.

kkonstantine commented:
Failed builds: 
[jdk8](https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/2023/)
[jdk11](https://builds.apache.org/job/kafka-pr-jdk11-scala2.13/6045/)
[jdk14](https://builds.apache.org/job/kafka-pr-jdk14-scala2.13/175/)

tombentley committed:
Fix broken test

tombentley commented:
@kkonstantine I should have run all the tests. Lesson learned, and sorry for wasting your time. Now fixed (and tested).

kkonstantine commented:
no problem @tombentley 
It's the VMs doing the heavy lifting from here :)

kkonstantine commented:
retest this please

kkonstantine commented:
Build results: 
jdk8 2 unrelated failures: 
`org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.testReplication`
`org.apache.kafka.streams.integration.QueryableStateIntegrationTest.shouldAllowConcurrentAccesses`
jdk11 1 unrelated failure: 
`org.apache.kafka.streams.integration.GlobalKTableIntegrationTest.shouldKStreamGlobalKTableJoin`
jdk14 green

kkonstantine reviewed (approved):
LGTM!

tombentley commented:
Thanks!

jherico commented on clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:
```diff
@@ -483,6 +483,7 @@ public void logUnused() {
                 resolvedOriginals.putAll(result.data());
             }
         }
+        providers.values().forEach(x -> Utils.closeQuietly(x, "config provider"));
```
if an exception occurs between line 477 and line 486, then the close function won't be called.  The solution would be to either create a type for `providers` that extends `Map<String, ConfigProvider>` and implements Autoclosable, or to simply put an explicit `try`/`finally` block here to ensure that the close function is called in every case.  That also implies that `instantiateConfigProviders` should be modified so that if an exception is thrown from inside it, any previously opened `ConfigProvider` instances are closed.

kkonstantine replied:
Good observations @jherico . I believe the latter approach, of closing all the instantiated providers if an exception occurs in both cases, would be the most straightforward fix. 

If you'd be interested in submitting a fix, that would be very welcome!
